src/code/mod.rs

```diff
@@ -0,0 +1,232 @@
+/// `CodeBuffer`s which will be intended according to their own indent, on top
+/// of the containing buffer's indent.
+pub struct CodeBuffer {
+    indent: Cow<'static, str>,
+    content: RefCell<Vec<CodeBufferContent>>,
+}
+
+impl CodeBuffer {
+    /// Creates a new `CodeBuffer` with no identation.
+    pub const fn new() -> Self {
+        Self::with_indent(Cow::Borrowed(""))
+    }
+
+    /// Adds a single newline character to this code buffer.
+    #[inline]
+    pub fn newline(&self) {
+        self.line(Cow::Borrowed(""))
+    }
+
+    /// Adds text into the buffer, followed by a new line.
+    pub fn line(&self, text: impl Into<Cow<'static, str>>) {
+        let mut content = self.content.borrow_mut();
+        content.push(CodeBufferContent::String(text.into(), true));
+    }
+
+    /// Adds text into the buffer, as-is.
+    pub fn text(&self, text: impl Into<Cow<'static, str>>) {
+        let mut content = self.content.borrow_mut();
+        content.push(CodeBufferContent::String(text.into(), false));
+    }
+
+    /// Creates a new indented sub-buffer at the current position.
+    #[inline]
+    pub fn indent(&self, indent: Cow<'static, str>) -> Rc<CodeBuffer> {
+        self.indent_with_options(IndentOptions {
+            indent,
+            leading: None,
+            trailing: None,
+            trailing_newline: false,
+        })
+    }
+
+    /// Creates a new indented sub-buffer at the current position.
+    pub fn indent_with_options(&self, options: IndentOptions) -> Rc<CodeBuffer> {
+        let mut content = self.content.borrow_mut();
+        if let Some(leading) = options.leading {
+            content.push(CodeBufferContent::String(leading, true));
+        }
+
+        let idx = content.len();
+        content.push(CodeBufferContent::Buffer(Rc::new(CodeBuffer::with_indent(
+            Cow::Owned(format!("{}{}", self.indent, options.indent)),
+        ))));
+
+        if let Some(trailing) = options.trailing {
+            content.push(CodeBufferContent::String(
+                trailing,
+                options.trailing_newline,
+            ));
+        } else if options.trailing_newline {
+            content.push(CodeBufferContent::String(Cow::Borrowed(""), true));
+        }
+
+        content[idx].as_buffer()
+    }
+
+    /// Creates a new un-indented sub-buffer at the current position.
+    #[inline]
+    pub fn section(&self, trailing_newline: bool) -> Rc<CodeBuffer> {
+        self.indent_with_options(IndentOptions {
+            indent: Cow::Borrowed(""),
+            leading: None,
+            trailing: None,
+            trailing_newline,
+        })
+    }
+
+    /// Writes the content of this `CodeBuffer` into the provided writer.
+    pub fn write(self, writer: &mut dyn io::Write) -> io::Result<()> {
+        self.inner_write(&mut IndentedWriter::new(writer))
+    }
+
+    fn inner_write(&self, writer: &mut IndentedWriter) -> io::Result<()> {
```

Comment:
In case you're wondering... I'd have preferred `inner_write` to also consume `self`, however this is not possible because we have to shove the sub-`CodeBuffer`s into `Rc` (reference counting smart pointers), which does not necessarily allow us to actually regain ownership of the underlying value (`Rc::try_unwrap`, as the name implies, only transfers ownership if the reference count is exactly `1`, which may not be the case). This is (in parts) why there is an `inner_write` API that is private here.